pkg/operators/sort/sort.go

```diff
@@ -0,0 +1,276 @@
+			v2, _ := f.Int32(j)
+			return (v1 < v2) != negate
+		}
+	case api.Kind_Int64:
+		return func(i, j datasource.Data) bool {
+			v1, _ := f.Int64(i)
+			v2, _ := f.Int64(j)
+			return (v1 < v2) != negate
+		}
+	case api.Kind_Uint8:
+		return func(i, j datasource.Data) bool {
+			v1, _ := f.Uint8(i)
+			v2, _ := f.Uint8(j)
+			return (v1 < v2) != negate
+		}
+	case api.Kind_Uint16:
+		return func(i, j datasource.Data) bool {
+			v1, _ := f.Uint16(i)
+			v2, _ := f.Uint16(j)
+			return (v1 < v2) != negate
+		}
+	case api.Kind_Uint32:
+		return func(i, j datasource.Data) bool {
+			v1, _ := f.Uint32(i)
+			v2, _ := f.Uint32(j)
+			return (v1 < v2) != negate
+		}
+	case api.Kind_Uint64:
+		return func(i, j datasource.Data) bool {
+			v1, _ := f.Uint64(i)
+			v2, _ := f.Uint64(j)
+			return (v1 < v2) != negate
+		}
+	case api.Kind_Float32:
+		return func(i, j datasource.Data) bool {
+			v1, _ := f.Float32(i)
+			v2, _ := f.Float32(j)
+			return (v1 < v2) != negate
+		}
+	case api.Kind_Float64:
+		return func(i, j datasource.Data) bool {
+			v1, _ := f.Float64(i)
+			v2, _ := f.Float64(j)
+			return (v1 < v2) != negate
+		}
+	case api.Kind_String, api.Kind_CString:
+		return func(i, j datasource.Data) bool {
+			v1, _ := f.String(i)
+			v2, _ := f.String(j)
+			if strings.Compare(v1, v2) < 0 {
+				return !negate
+			}
+			return negate
+		}
+	default:
+		return nil
+	}
+}
+
+func (s *sortOperatorInstance) init(gadgetCtx operators.GadgetContext) error {
+	s.sorters = make(map[datasource.DataSource][]func(i datasource.Data, j datasource.Data) bool)
+	dsSorts := make(map[string][]string)
+	for _, srt := range strings.Split(s.sortBy, ";") {
+		dsFields := strings.Split(srt, ":")
+		dsName := ""
+		fieldList := dsFields[0]
+		if len(dsFields) == 2 {
+			dsName = dsFields[0]
+			fieldList = dsFields[1]
+		}
+		fields := strings.Split(fieldList, ",")
+		if len(fields) > 2 || fields[0] != "" {
+			dsSorts[dsName] = fields
+		}
+	}
+
+	// Check edge cases
+	dsSpecific := true
+	if _, ok := dsSorts[""]; ok {
+		if len(dsSorts) > 1 {
+			return fmt.Errorf("mixing sorting rules with and without specifying data source")
+		}
+		dsSpecific = false
+	}
+
+	for _, ds := range gadgetCtx.GetDataSources() {
+		// Can only work on arrays
+		if ds.Type() != datasource.TypeArray {
+			continue
+		}
+
+		sortFields := dsSorts[ds.Name()]
+		if !dsSpecific {
+			sortFields = dsSorts[""]
+		}
+
+		if len(sortFields) == 0 {
+			continue
+		}
+
+		var sortFuncs []func(i, j datasource.Data) bool
+		for _, fieldName := range sortFields {
+			fieldName, negate := strings.CutPrefix(fieldName, "-")
+
+			field := ds.GetField(fieldName)
+			if field == nil {
+				return fmt.Errorf("field %s not found", fieldName)
+			}
+
+			cmp := getCompareFunc(field, negate)
+			if cmp == nil {
+				return fmt.Errorf("field %s cannot be used for sorting", fieldName)
+			}
+			sortFuncs = append(sortFuncs, cmp)
+		}
+		if len(sortFuncs) == 0 {
+			continue
+		}
+
+		slices.Reverse(sortFuncs)
```

Comment:
> (it seems to me it's first sorting by pid and then by comm)

For me this is the intuitively correct behavior: instead of seeing the parameter as a "sort by comm and then by -pid", I read it as "the result should be sorted by comm and within that sorting by -pid". That's why I'm reversing the sortFuncs - it sorts lower priorities first and then uses a stable sort to keep the lower priority sorting when doing the higher priority ones.

This is also how SQL handles sorting of multiple columns (e.g. `ORDER BY comm, pid DESC`).

Do you have other examples that behave differently?

Reply:
I think we're on the same page but using different words. I don't know why, but in the previous iteration I think I have to remove that line to make it work right, now it works fine (after rebasing on top of #3057 and importing the operator on ig). 

I think we both agree the following output is the correct one:

```bash
$ sudo -E ig run snapshot_process:latest --sort comm,-pid  --fields=comm,pid
INFO[0000] Experimental features enabled
WARN[0000] Ignoring runtime "cri-o" with non-existent socketPath "/run/crio/crio.sock"
WARN[0000] Ignoring runtime "podman" with non-existent socketPath "/run/podman/podman.sock"
COMM             PID
containerd       183790
containerd-shim  186795
containerd-shim  186304
containerd-shim  185989
containerd-shim  185793
containerd-shim  185703
containerd-shim  185655
containerd-shim  185587
containerd-shim  185546
containerd-shim  185380
containerd-shim  185249
containerd-shim  185242
containerd-shim  185236
containerd-shim  185192
containerd-shim  185100
containerd-shim  185099
containerd-shim  185035
cri-dockerd      184401
dockerd          184048
kubelet          184666
sshd             183077
systemd          182871
systemd-journal  183031
```

Reply:
> I think we're on the same page but using different words.

Haha, yeah, sorry - I didn't look closely enough at the comm sorting in your example.

> I think we both agree the following output is the correct one

Yes!

The reason it was buggy before was that it didn't really use stable sorting.